src/www/app/outline_server.ts

```diff
@@ -223,45 +219,88 @@ export class OutlineServerRepository implements ServerRepository {
   private loadServers() {
+    if (this.storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY)) {
+      console.debug('server storage migrated to V1');
+      this.loadServersV1();
+      return;
+    }
+    this.loadServersV0();
+  }
+
+  private loadServersV0() {
     this.serverById = new Map<string, OutlineServer>();
-    const serversJson = this.storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY);
+    const serversJson = this.storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY_V0);
     if (!serversJson) {
-      console.debug(`no servers found in storage`);
+      console.debug(`no V0 servers found in storage`);
       return;
     }
-    let configByAccessKey: AccessKeyById = {};
+    let configById: ServersStorageV0 = {};
     try {
-      configByAccessKey = JSON.parse(serversJson);
+      configById = JSON.parse(serversJson);
     } catch (e) {
-      throw new Error(`could not parse saved servers: ${e.message}`);
+      throw new Error(`could not parse saved V0 servers: ${e.message}`);
     }
-    for (const serverId in configByAccessKey) {
-      if (configByAccessKey.hasOwnProperty(serverId)) {
-        const accessKey = configByAccessKey[serverId];
+    for (const serverId in configById) {
+      if (configById.hasOwnProperty(serverId)) {
+        const config = configById[serverId];
         try {
-          const config = accessKeyToShadowsocksConfig(accessKey);
-          const server = this.createServer(serverId, config, this.eventQueue);
-          if (!OutlineServer.isServerCipherSupported(config.method)) {
-            server.errorMessageId = 'unsupported-cipher';
-          }
-          this.serverById.set(serverId, server);
+          this.loadServer(serverId, shadowsocksConfigToAccessKey(config), config);
         } catch (e) {
           // Don't propagate so other stored servers can be created.
           console.error(e);
         }
       }
     }
   }
+
+  private loadServersV1() {
+    this.serverById = new Map<string, OutlineServer>();
+    const serversStorageJson = this.storage.getItem(OutlineServerRepository.SERVERS_STORAGE_KEY);
+    if (!serversStorageJson) {
+      console.debug(`no servers found in storage`);
+      return;
+    }
+    let serversJson: ServersStorageV1 = [];
+    try {
+      serversJson = JSON.parse(serversStorageJson);
+    } catch (e) {
+      throw new Error(`could not parse saved servers: ${e.message}`);
+    }
+    for (const serverJson of serversJson) {
+      try {
+        const config = accessKeyToShadowsocksConfig(serverJson.accessKey);
+        if (serverJson.name) {
+          config.name = serverJson.name;
+        }
+        this.loadServer(serverJson.id, serverJson.accessKey, config);
+      } catch (e) {
+        // Don't propagate so other stored servers can be created.
+        console.error(e);
+      }
+    }
+  }
+
+  private loadServer(serverId: string, accessKey: string, config: ShadowsocksConfig) {
```

Comment:
Change loadServer to take one parameter: OutlineServerJson.
It will simplify the loadServer methods.
Also, if you can't load the server from the OutlineServerJson, something is wrong. That will make it clearer what is needed to load a server.

Reply:
Done in 3e18673.